PORT-14552 - Docs for Github ocean integration (#2494)
port-labs/port-docs: docs/build-your-software-catalog/sync-data-to-catalog/git/github-ocean/github-ocean.md

---
sidebar_class_name: hidden
---

import Tabs from "@theme/Tabs"
import TabItem from "@theme/TabItem"
import GitHubResources from './\_github_exporter_supported_resources.mdx'

# GitHub Self-Hosted (Beta)

Port's GitHub self-hosted integration allows you to model GitHub resources in your software catalog and ingest data into them.

## Overview

Here's what you can do with the GitHub integration:

- Map and organize your desired GitHub resources and their metadata in Port (see supported resources below).
- Watch for GitHub object changes (create/update/delete) in real-time, and automatically apply the changes to your software catalog.
- Manage Port entities using GitOps.

### Supported resources

The resources that can be ingested from GitHub into Port are listed below.
It is possible to reference any field that appears in the API responses linked below in the mapping configuration.

 <GitHubResources/>

## Setup

> To install Port's GitHub integration, see [our dedicated installation guides](./installation).

Reviewer: ```suggestion
To install the integration, see the [installation page](./installation).
```

## Configuration

Port integrations use a [YAML mapping block](/build-your-software-catalog/customize-integrations/configure-mapping#configuration-structure) to ingest data from the third-party api into Port.

The mapping makes use of the [JQ JSON processor](https://stedolan.github.io/jq/manual/) to select, modify, concatenate, transform and perform other operations on existing fields and values from the integration API.

> To ingest GitHub objects, use one of the following methods:

Reviewer: there is only one method (UI), so no need for tabs

<Tabs queryString="method">

<TabItem label="Using Port's UI" value="port">

To manage your GitHub integration configuration using Port:

1. Go to the [data sources](https://app.getport.io/settings/data-sources) page of your portal.
2. Under `Exporters`, click on the installed integration.
3. A window will open containing the default YAML configuration of your GitHub integration.
4. Here you can modify the configuration to suit your needs, by adding/removing entries.
5. When finished, click `resync` to apply any changes.

Using this method applies the configuration to all repositories that the GitHub app has permissions to.

When configuring the integration **using Port**, the YAML configuration is global, allowing you to specify mappings for multiple Port blueprints.

</TabItem>

</Tabs>

> ### Default mapping configuration

Author: Having it here makes it consistent with other integrations - like the existing Github app, Gitlab-v2, and Bitbucket Cloud.

> ### Default mapping configuration

Reviewer 2: It's just that maybe having the default mapping in the introduction page for the integration might be too soon. IMO having a dedicated page for example mappings + mapping configurations  might make more sense, but if we don't want to make big changes then there's no reason.

    port:
      entity:
        mappings:
          identifier: ".head.repo.name + (.id|tostring)"
          title: ".title"
          blueprint: '"githubPullRequest"'
          properties:
            creator: ".user.login"
            assignees: "[.assignees[].login]"
            reviewers: "[.requested_reviewers[].login]"
            status: ".state"
            closedAt: ".closed_at"
            updatedAt: ".updated_at"
            mergedAt: ".merged_at"
            createdAt: ".created_at"
            prNumber: ".id"
            link: ".html_url"
            leadTimeHours: >-
                (.created_at as $createdAt | .merged_at as $mergedAt |
                ($createdAt | sub("\\..*Z$"; "Z") | strptime("%Y-%m-%dT%H:%M:%SZ") | mktime) as $createdTimestamp |
                ($mergedAt | if . == null then null else sub("\\..*Z$"; "Z") |
                strptime("%Y-%m-%dT%H:%M:%SZ") | mktime end) as $mergedTimestamp |
                if $mergedTimestamp == null then null else
                (((($mergedTimestamp - $createdTimestamp) / 3600) * 100 | floor) / 100) end)
          relations:
            repository: .__repository
```

</details>

## Capabilities

> ### Ingesting Git objects

Reviewer: ```suggestion
### Ingest Git objects
```

itemsToParse: .content | if type== "object" then [.] else . end
```
:::

#### Ingest raw file content

If you need to ingest the raw content of a file without parsing it, you can use the `skipParsing` key in your file selector.  
This is useful when you want to store the file content as a string or YAML property.  

When `skipParsing` is set to `true`, the file content will be kept in its original string format instead of being parsed into a JSON/YAML object.

Here's an example that ingests the raw content of a `values.yaml` file into the `content` property of a `file` entity:

```yaml
resources:
  - kind: file
    selector:
      query: 'true'
      files:
        - path: values.yaml
          skipParsing: true
    port:
      entity:
        mappings:
          identifier: >-
            .repository.name + "-values"
          blueprint: '"file"'
          properties:
            content: .content
```

#### Limitations

- Currently only files up to 1MB in size are supported.
- Only JSON and YAML formats are automatically parsed.  
  Other file formats can be ingested as raw files, however, some special characters in the file (such as `\n`) may be processed and not preserved.

> ## Permissions

Author: @hadar-co I'm wondering if this isn't redundant information, there is a `required permissions` in the installation page as well, I think that is a much better place to have it. My instinct is to remove this entirely.
